docs/design/audit-v2.md

```diff
@@ -0,0 +1,74 @@
+- One loop will select based on stripes.
+- The second loop will select based on nodes.
+
+After selection, the rest of the auditing process will occur the same way (picking a segment, picking a stripe, downloading all erasure shares associated with that stripe).
+The chances of selecting the same stripe are rare, but should be accounted for.
+
+With both loops, we should have auditing that occurs statistically uniform across both nodes and bytes.
+
+### **Selection via Reservoir Sampling**
+
+Reservoir sampling: a family of algorithms for randomly choosing a sample of items with uniform probability from a stream of data (of unknown size).
+
+During uploads or during garbage collection bloom filter generation, we're going through all segments and creating filters per node.
+This is a good opportunity to insert reservoir sampling for auditing purposes.
+
+We will have little reservoirs of segments for all of the nodes.
+We could generate 5 random segments per node, then randomly select some segments per node.
+By choosing nodes out of the reservoirs, we would receive a random sample.
+Unvetted nodes would get the same number of audits as vetted nodes.
+
+However, one problem with piggybacking off of existing garbage collection bloom filter generation is that bloom filter generation occurs every seven days.
+
+The gc observer subscribes to the metainfo loop every week, but we could make an observer that subscribes more often and gets a stream as often as we want, so it doesn't have to be part of gc.
+However, this could potentially create a lot more db transactions, which we want to minimize.
+[See Open Issue 1.]
+
+## Rationale
+
+An initial idea for implementation was to sort the nodes table for nodes with least amount of audits, then select one node randomly within that low amount of audits.
+However, we decided it may not be necessary to keep track of how many audits per storage node if we're able to randomly select across nodes.
+
+We were also considering implementing a reverse way of looking up segments or pieces by node IDs.
+e.g. a table where each row is a node ID and an encrypted metainfo path.
+Every time a segment is committed or deleted, that table (and every node) gets updated.
+The advantage is that it can super simplify the garbage collection process, but complexify upload and download.
+
+## Implementation
+
+1. Nail down open issue 1 and identify where reservoir sampling should occur.
+2. Create new part of audit service that loops over nodes.
+3. Add new methods to this loop that allow for selection from reservoir.
+4. Coordinate both selection loops to use same "audit path" for verification.
+
+## Open issues
+
+1. Location of reservoir sampling.
+    - Where do we have this happen?
+    - We want uploads to be performant with minimal db transactions, but we know that audits need to happen very frequently.
+
+2. How to coordinate both loops running simultaneously/using same "audit path" once data is selected for audit?
```

Comment:
I think this could be reworded to something like "Should we run both audit selection processes within the same loop or in separate loops?"

Reply:
I think that's asking a different question though, because I'm definitely anticipating two loops, but I guess I was thinking they wouldn't be able to do the regular verifying functions at the same time. But actually now that I think about it, that shouldn't be an issue. I might get rid of this question!

Reply:
Or I guess I'll just add the question you suggested.

Reply:
Yeah, I think if there are two separate loops, there is no coordination to be done, except maaaybe avoiding re-audits of the same segment